### PR TITLE
[Operator] Add Conv3d forward function

### PR DESCRIPTION
**Pull request by Gxiandy**
### PR Category

Operator 
### Type of Change

New Feature
### Description

Add Conv3d forward function and related tests
### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### Performance

Operator: conv3d  Performance Test (dtype=torch.float16, mode=cuda,level=core)
|Status|       Torch Latency (ms) |   Gems Latency (ms) |        Gems Speedup |         Size Detail|
|----------|---------------------------|--------------------|--------------|------------------------|
|SUCCESS|               6.790144   |         6.793216              | 1.000 |         {'input': torch.Size([104, 16, 32, 32, 32]), 'weight': torch.Size([32, 16, 4, 4, 4]), 'bias': None, 'groups': 1, 'stride': 1, 'padding': 0}|
|SUCCESS  |             4.712448     |       4.705280              | 1.002   |       {'input': torch.Size([64, 32, 18, 180, 18]), 'weight': torch.Size([32, 32, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS    |           0.943104       |     0.941056            |   1.002     |     {'input': torch.Size([4, 32, 110, 110, 10]), 'weight': torch.Size([64, 32, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS      |         2.338816         |   2.342912          |     0.998       |   {'input': torch.Size([4, 64, 110, 110, 10]), 'weight': torch.Size([16, 64, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS        |       0.208896           | 0.208896        |       1.000         | {'input': torch.Size([16, 32, 120, 12, 12]), 'weight': torch.Size([24, 32, 3, 3, 3]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS          |     7.403520           | 7.422976      |         0.997          |{'input': torch.Size([16, 32, 240, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 1, 'padding': 1}|
|SUCCESS            |   0.224256            |0.224256    |           1.000          |{'input': torch.Size([16, 32, 24, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 2, 'padding': 2}|
|SUCCESS              | 0.933888            |0.931840  |             1.002        |  {'input': torch.Size([16, 32, 24, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 1, 'padding': 2}|

Operator: conv3d  Performance Test (dtype=torch.float32, mode=cuda,level=core)
|Status     |  Torch Latency (ms|)    Gems Latency (ms)     |    Gems Speedup   |       Size Detail|
|-----------|--------------------|---------------------------|-------------------|-----------------|
|SUCCESS     |         35.387390  |         35.483646         |      0.997       |   {'input': torch.Size([104, 16, 32, 32, 32]), 'weight': torch.Size([32, 16, 4, 4, 4]), 'bias': None, 'groups': 1, 'stride': 1, 'padding': 0}|
|SUCCESS      |        20.566015   |        21.796864          |     0.944       |   {'input': torch.Size([64, 32, 18, 180, 18]), 'weight': torch.Size([32, 32, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS       |        5.250048    |        5.249024           |    1.000       |   {'input': torch.Size([4, 32, 110, 110, 10]), 'weight': torch.Size([64, 32, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS        |       5.134336     |       5.169152            |   0.993       |   {'input': torch.Size([4, 64, 110, 110, 10]), 'weight': torch.Size([16, 64, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS         |      0.425984      |      0.431104             |  0.988       |   {'input': torch.Size([16, 32, 120, 12, 12]), 'weight': torch.Size([24, 32, 3, 3, 3]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS          |   103.790588       |   103.080963             |  1.007       |   {'input': torch.Size([16, 32, 240, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 1, 'padding': 1}|
|SUCCESS           |    0.407552        |    0.414720             |  0.983       |   {'input': torch.Size([16, 32, 24, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 2, 'padding': 2}|
|SUCCESS            |  10.667008         |  10.584064             |  1.008       |   {'input': torch.Size([16, 32, 24, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 1, 'padding': 2}|

Operator: conv3d  Performance Test (dtype=torch.bfloat16, mode=cuda,level=core)
|Status|       Torch Latency (ms)  |  Gems Latency (ms) |        Gems Speedup|          Size Detail|
|-------|---------------------------|--------------------|--------------------|---------------------|
|SUCCESS|              11.389952     |      11.388928     |          1.000     |     {'input': torch.Size([104, 16, 32, 32, 32]), 'weight': torch.Size([32, 16, 4, 4, 4]), 'bias': None, 'groups': 1, 'stride': 1, 'padding': 0}|
|SUCCESS|               4.552704      |      4.488192      |         1.014      |    {'input': torch.Size([64, 32, 18, 180, 18]), 'weight': torch.Size([32, 32, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS|               0.941056       |     0.943104       |        0.998       |   {'input': torch.Size([4, 32, 110, 110, 10]), 'weight': torch.Size([64, 32, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS|               3.181568        |    3.180544        |       1.000        |  {'input': torch.Size([4, 64, 110, 110, 10]), 'weight': torch.Size([16, 64, 5, 5, 5]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS|               0.256000         |   0.256000         |      1.000        |  {'input': torch.Size([16, 32, 120, 12, 12]), 'weight': torch.Size([24, 32, 3, 3, 3]), 'bias': None, 'groups': 1, 'stride': 2, 'padding': 1}|
|SUCCESS|               7.410688          |  7.752704          |     0.956        |  {'input': torch.Size([16, 32, 240, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 1, 'padding': 1}|
|SUCCESS|               0.222208           | 0.223232           |    0.995        |  {'input': torch.Size([16, 32, 24, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 2, 'padding': 2}|
|SUCCESS|               0.931840            |0.934912            |   0.997        |  {'input': torch.Size([16, 32, 24, 24, 24]), 'weight': torch.Size([24, 16, 3, 3, 3]), 'bias': None, 'groups': 2, 'stride': 1, 'padding': 2}|

### CODE REVIEW

**Comment by StrongSpoon** [comment]
impressive performance! we will review soon.

File: benchmark/core_shapes.yaml [hunk 1]
```diff
@@ -178,3 +178,14 @@ AttentionBenchmark:
...
+Conv3dBenchmark:
+  shapes:
+  - [104, 16, 32, 32, 32, 32, 4, 4, 4, 1, 0, 1]
+  - [64, 32, 18, 180, 18, 32, 5, 5, 5, 2, 1, 1]
+  - [4, 32, 110, 110, 10, 64, 5, 5, 5, 2, 1, 1]
+  - [4, 64, 110, 110, 10, 16, 5, 5, 5, 2, 1, 1]
+  - [16, 32, 120, 12, 12, 24, 3, 3, 3, 2, 1, 1]
+  - [16, 32, 240, 24, 24, 24, 3, 3, 3, 1, 1, 2]
+  - [16, 32, 24, 24, 24, 24, 3, 3, 3, 2, 2, 2]
+  - [16, 32, 24, 24, 24, 24, 3, 3, 3, 1, 2, 2]
```
Comment (StrongSpoon) [inline benchmark/core_shapes.yaml:191]: usually we suggest setting 5 shapes for core mode.

File: benchmark/test_reduction_perf.py [hunk 2]
```diff
@@ -186,3 +187,67 @@ def count_nonzero_input_fn(shape, dtype, device):
...
+            for d in [128 * i for i in range(5, 9)]
+            for h in [128 * i for i in range(5, 9)]
+            for w in [128 * i for i in range(5, 9)]
+            for o_c in [128 * i for i in range(8, 9)]
+            for k_d in [32 * i for i in range(1, 4)]
+            for k_h in [32 * i for i in range(1, 4)]
+            for k_w in [32 * i for i in range(1, 4)]
+            for stride in [1, (2, 2, 2), (3, 3, 3)]
+            for padding in [0, (1, 1, 1), (0, 1, 2)]
+            for groups in [1, 2, 4, 8]
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py:209]: personally I think the number of test cases is too large. pick 20 shapes from classic networks and the performance data is convincing enough.

File: benchmark/test_reduction_perf.py [hunk 3]
```diff
@@ -186,3 +187,67 @@ def count_nonzero_input_fn(shape, dtype, device):
...
+            "groups": groups,
+            "stride": stride,
+            "padding": padding,
+        },
+
+    torch.backends.cudnn.allow_tf32 = False
+    bench = Conv3dBenchmark(
+        input_fn=conv3d_input_fn,
+        op_name="conv3d",
+        torch_op=torch.nn.functional.conv3d,
```
Comment (StrongSpoon) [inline benchmark/test_reduction_perf.py:250]: since your implementation of conv3d is not registered into aten library, the function called in benchmark is still torch op. please add registration in src/flag_gems/__init__.py and update the benchmark results.

File: benchmark/test_special_perf.py [hunk 4]
```diff
@@ -264,7 +264,7 @@ def set_more_shapes(self):
 
 @pytest.mark.conv2d
 def test_perf_conv2d():
-    def conv2d_input_fn(shape, dtype, device):
+    def conv3d_input_fn(shape, dtype, device):
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py:267]: it's not time to enable benchmark of conv2d right now. so does conv3d. I'll mark it as skip.

File: tests/test_reduction_ops.py [hunk 5]
```diff
@@ -874,3 +874,53 @@ def test_accuracy_depthwise2d(
...
+@pytest.mark.parametrize("dilations", [1, (2, 2, 2), (1, 2, 3)])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.float32])
+def test_accuracy_conv3d(
+    shape_input, shape_weight, groups, strides, paddings, dilations, dtype
+):
+    inp = torch.randn(
+        shape_input, dtype=dtype, device=flag_gems.device, requires_grad=True
+    )
+    ref_inp = to_reference(inp, True)
+    torch.backends.cudnn.allow_tf32 = True
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py:901]: usually we set allow_tf32 as False since the precision of tf32 is not satisfying.

File: tests/test_reduction_ops.py [hunk 6]
```diff
@@ -874,3 +874,53 @@ def test_accuracy_depthwise2d(
...
+
+    ref_out = torch.nn.functional.conv3d(
+        ref_inp,
+        ref_weight,
+        bias=None,
+        groups=groups,
+        stride=strides,
+        padding=paddings,
+        dilation=dilations,
+    ).to(dtype)
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py:913]: gems_assert_close will cast the reference tensor to dtype. you don't need to do this again.

File: src/flag_gems/ops/conv3d.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+        # Initialize the output value
+        value = tl.zeros((), dtype=tl.float32)
+
+        # Perform convolution
+        group_id = k_idx // K_per_group
+        for t in range(T):
+            for r in range(R):
+                for s in range(S):
+                    for c in range(C_per_group):
```
Comment (StrongSpoon) [inline src/flag_gems/ops/conv3d.py:81]: using multiple layers of loop might not be a good idea to compute convolution. try loading tensors with high-dimension indexes and using tl.dot primitive.

**Comment by StrongSpoon** [comment]
this pr hasn't received a response for too long. I'll close it.
